src/charm.py

```diff
@@ -108,6 +107,24 @@ def __init__(self, *args):
         self.framework.observe(self.db.on.standby_changed, self._on_standby_changed)
         self.framework.observe(self.on.db_master_available, self.configure_pod)
 
+    def _on_grant_admin_role_action(self, event):
+        """Handle the grant-admin-role action."""
+        user = event.params["user"]
+        cmd = ["/mattermost/bin/mattermost", "roles", "system_admin", user]
+        granted = subprocess.run(cmd, capture_output=True)
+        if granted.returncode != 0:
+            event.fail(
+                "Failed to run '{}'. Output was:\n{}".format(
+                    " ".join(cmd), granted.stderr.decode("utf-8")
+                )
+            )
+        else:
+            msg = (
+                "Ran grant-admin-role for user '{}'. They will need to log out and log back in "
+                "to Mattermost to fully receive their permissions upgrade.".format(user)
```

Comment:
f-string?

Reply:
Apart from the licence/license change and the removal of the functions put in environment.py, I haven't changed the charm.py.
I'll consider this later